src/app/grapheneos/geocoder/NominatimGeocoder.kt

```diff
@@ -0,0 +1,177 @@
+        upperRightLongitude: Double,
+        maxResults: Int,
+        preferredLocale: Locale
+    ): List<Address> {
+        val urlSuffix =
+            "/search?format=geocodejson&q=$locationName&limit=$maxResults&addressdetails=1&extratags=1".let {
+                // add bounding box parameter if it's set for this request
+                if (lowerLeftLatitude == 0.0 && lowerLeftLongitude == 0.0 && upperRightLatitude == 0.0 && upperRightLongitude == 0.0) {
+                    it
+                } else {
+                    "$it&viewbox=$lowerLeftLongitude,$lowerLeftLatitude,$upperRightLongitude,$upperRightLatitude&bounded=1"
+                }
+            }
+        return fetch(urlSuffix, maxResults, preferredLocale)
+    }
+
+    @Throws(IOException::class)
+    override fun reverseGeocode(
+        latitude: Double,
+        longitude: Double,
+        maxResults: Int,
+        preferredLocale: Locale,
+    ): List<Address> {
+        val urlSuffix =
+            // Nominatim doesn't support returning more than 1 reverse geocoding result
+            "/reverse?format=geocodejson&lat=${latitude}&lon=${longitude}&zoom=18&addressdetails=1&extratags=1"
+        return fetch(urlSuffix, maxResults, preferredLocale)
+    }
+
+    @Throws(IOException::class)
+    private fun fetch(
+        urlSuffix: String,
+        expectedMaxResults: Int,
+        preferredLocale: Locale
+    ): List<Address> {
+        val (baseUrl, enforceModernTls) = getServerBaseUrl()
+        val url = URL("$baseUrl$urlSuffix")
+
+        val connection = url.openConnection() as HttpsURLConnection
+        val responseBytes = try {
+            if (enforceModernTls) {
+                connection.sslSocketFactory = tlsSocketFactory
+            }
+            connection.requestMethod = "GET"
+            connection.setRequestProperty("Accept-Language", preferredLocale.toLanguageTag())
+            connection.connectTimeout = 10_000
+            connection.readTimeout = 10_000
+
+            val responseCode = connection.responseCode
+            if (responseCode != HttpsURLConnection.HTTP_OK) {
+                throw IOException("non-200 response code: $responseCode")
+            }
+            connection.inputStream.use { inputStream ->
+                inputStream.readAllBytes()
+            }
+        } finally {
+            connection.disconnect()
+        }
+
+        val responseString = responseBytes.decodeToString()
+
+        // occurs when reverse geocode coordinates are in the middle of the ocean, presumably
+        // because there were no results
+        if (responseString == "{\"error\":\"Unable to geocode\"}") {
+            return emptyList()
+        }
+
+        val response = try {
+            val json = Json { ignoreUnknownKeys = true }
+            json.decodeFromString<GeocodeJson>(responseString)
+        } catch (e: SerializationException) {
+            throw IOException("failed to deserialize response: $e")
```

Comment:
Exceptions aren't meant to be stringified like this, there's `cause` parameter for that purpose:
IOException("failed to deserialize response", e).

Same applies to the exception 2 lines below.